File: common/src/main/java/com/sparta/ditto/common/response/ApiResponse.java
package com.sparta.ditto.common.response;

import com.fasterxml.jackson.annotation.JsonInclude;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Getter;
import org.springframework.http.HttpStatus;

import java.util.List;

@Getter
@AllArgsConstructor(access = AccessLevel.PRIVATE)
@JsonInclude(JsonInclude.Include.NON_NULL)
public class ApiResponse<T> {

    private final int status;
    private final String code;
    private final String message;
    private final T data;
    private final List<String> errors;

    public static <T> ApiResponse<T> success(T data) {
        return new ApiResponse<>(HttpStatus.OK.value(), null, "SUCCESS", data, null);
    }

    public static <T> ApiResponse<T> success() {
        return new ApiResponse<>(HttpStatus.OK.value(), null, "SUCCESS", null, null);
    }

    public static <T> ApiResponse<T> created(T data) {
        return new ApiResponse<>(HttpStatus.CREATED.value(), null, "CREATED", data, null);
    }

    public static <T> ApiResponse<T> updated(T data) {
        return new ApiResponse<>(HttpStatus.OK.value(), null, "UPDATED", data, null);
    }

    public static <T> ApiResponse<Void> deleted() {
        return new ApiResponse<>(HttpStatus.OK.value(), null, "DELETED", null, null);
    }

    public static <T> ApiResponse<Void> accepted() {
        return new ApiResponse<>(HttpStatus.ACCEPTED.value(), null, "ACCEPTED", null, null);
    }

    public static <T> ApiResponse<T> error(int status, String code, String message) {
        return new ApiResponse<>(status, code, message, null, null);
    }

    public static <T> ApiResponse<T> error(int status, String code, String message, List<String> errors) {
        return new ApiResponse<>(status, code, message, null, errors);
    }

    public static <T> ApiResponse<T> error(HttpStatus status, String code, String message) {
        return new ApiResponse<>(status.value(), code, message, null, null);
    }
}

File: embedding_service/app/common/response/api_response.py
from typing import Generic, TypeVar
from pydantic import BaseModel

T = TypeVar("T")


class ApiResponse(BaseModel, Generic[T]):
    status: int
    code: str | None = None
    message: str
    data: T | None = None
    errors: list[str] | None = None

    @classmethod
    def success(cls, data: T) -> "ApiResponse[T]":
        return cls(status=200, message="SUCCESS", data=data)

    @classmethod
    def success_no_content(cls) -> "ApiResponse[None]":
        return cls(status=200, message="SUCCESS")

    @classmethod
    def created(cls, data: T) -> "ApiResponse[T]":
        return cls(status=201, message="CREATED", data=data)

    @classmethod
    def updated(cls, data: T) -> "ApiResponse[T]":
        return cls(status=200, message="UPDATED", data=data)

    @classmethod
    def deleted(cls) -> "ApiResponse[None]":
        return cls(status=200, message="DELETED")

    @classmethod
    def accepted(cls) -> "ApiResponse[None]":
        return cls(status=202, message="ACCEPTED")

    @classmethod
    def error(cls, status: int, code: str, message: str, errors: list[str] | None = None) -> "ApiResponse[None]":
        return cls(status=status, code=code, message=message, errors=errors)
